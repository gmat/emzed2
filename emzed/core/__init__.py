from data.Table import Table
