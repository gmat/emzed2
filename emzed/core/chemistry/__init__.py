from elements import Elements, MonoIsotopicElements
from molecular_formula import MolecularFormula 
from tools import *
