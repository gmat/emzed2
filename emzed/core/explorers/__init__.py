from peakmap_explorer import inspectPeakMap, PeakMapExplorer
from table_explorer import inspect, TableExplorer
