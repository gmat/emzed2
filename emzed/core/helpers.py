import requests
def get_json(url):
    return requests.get(url, headers=dict(Accept="application/json"))
