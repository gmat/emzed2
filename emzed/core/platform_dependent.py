import os, sys
import functools


def userShellFolderKey():
    import _winreg
    key =_winreg.OpenKey(_winreg.HKEY_CURRENT_USER,
                        "Software\\Microsoft\\Windows\\CurrentVersion"
                        "\\Explorer\\User Shell Folders")
    return key

def _query(key, subKey):
    import _winreg
    val, _ = _winreg.QueryValueEx(key, subKey)
    return _winreg.ExpandEnvironmentStrings(val)

def linuxdefault(path):
    def wrapper(fun, path=path):
        @functools.wraps(fun)
        def new_fun():
            if sys.platform == "win32":
                return fun()
            else:
                return path
        return new_fun
    return wrapper


@linuxdefault(os.environ.get("HOME"))
def getDocumentFolder():
    key = userShellFolderKey()
    return _query(key, "Personal")

@linuxdefault(os.environ.get("HOME"))
def getAppDataFolder():
    key = userShellFolderKey()
    return _query(key, "AppData")

@linuxdefault(os.environ.get("HOME"))
def getLocalAppDataFolder():
    key = userShellFolderKey()
    return _query(key, "Local AppData")

def getEmzedFolder():
    if sys.platform == "win32":
        return os.path.join(getAppDataFolder(), "emzed")
    else:
        return os.path.join(getAppDataFolder(), ".emzed")


def getDataHome():
    dataHome = os.path.join(getDocumentFolder(), "emzed_files")
    return dataHome
