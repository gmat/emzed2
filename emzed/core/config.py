import pdb
#encoding: latin-1

# keep namespace clean:

import guidata.dataset.datatypes as _dt
import guidata.dataset.dataitems as _di

_is_expert = _dt.ValueProp(False)

class _UserConfig(object):


    class Parameters(_dt.DataSet):

        g1 = _dt.BeginGroup("User Settings")

        user_name = _di.StringItem("Full Name", notempty=True)
        user_email = _di.StringItem("Email Adress", notempty=True)
        user_url = _di.StringItem("Website URL")

        _g1 = _dt.EndGroup("User Settings")

        g2 = _dt.BeginGroup("Webservice Settings")

        metlin_token  = _di.StringItem("Metlin Token")

        _g2 = _dt.EndGroup("Webservice Settings")

        g3 = _dt.BeginGroup("Emzed Store User Account")

        emzed_store_user = _di.StringItem("User Name")
        emzed_store_password = _di.StringItem("User Password")

        _g3 = _dt.EndGroup("Emzed Store Settings")

        g4 = _dt.BeginGroup("Emzed Store Expert Settings")
        enable_expert_settings = _di.BoolItem("Enable Settings").set_prop("display",
                store=_is_expert)
        emzed_store_url = _di.StringItem("Emzed Store URL").set_prop("display", active=_is_expert)
        emzed_store_index_url = _di.StringItem("Emzed Store Index URL").set_prop("display", active=_is_expert)
        pypi_url = _di.StringItem("PyPi URL").set_prop("display", active=_is_expert)

        _g4 = _dt.EndGroup("Expert Settings")

    def __init__(self, *a, **kw):
        self.parameters = _UserConfig.Parameters()
        if "_no_load" not in kw:
            loaded = self.load()
            if not loaded:
                self.set_defaults()
        else:
            self.set_defaults()

    def get(self, key):
        val = getattr(self.parameters, key)
        if isinstance(val, unicode):
            val = val.encode("latin-1")
        return val

    def set_(self, key, value):
        return setattr(self.parameters, key, value)

    def get_url(self, key):
        return self.get(key).rstrip("/") + "/"

    def store(self, path=None):

        # path UserConfig for not writing to ~/.config/.none.ini as default:
        import guidata.userconfig
        __save = guidata.userconfig.UserConfig.__save
        try:
            guidata.userconfig.UserConfig.__save = lambda self: None
            if path is None:
                path = self.config_file_path()
            cf = guidata.userconfig.UserConfig(dict())
            self.parameters.write_config(cf, "emzed", "")
            with open(path, "wt") as fp:
                cf.write(fp)
        finally:
            # undo patch
            guidata.userconfig.UserConfig.__save = __save


    def load(self, path=None):
        import os
        import guidata.userconfig
        cf = guidata.userconfig.UserConfig(dict())
        if path is None:
            path = self.config_file_path()
        if os.path.exists(path):
            with open(path, "rt") as fp:
                try:
                    cf.readfp(fp)
                    self.parameters.read_config(cf, "emzed", "")
                    return True
                except:
                    pass
        return False

    def edit(self):
        self.parameters.edit()
        self.store()

    def config_file_path(self):
        import os
        from emzed.core.platform_dependent import getEmzedFolder
        return os.path.join(getEmzedFolder(), "config.ini")

    def set_defaults(self):
        self.parameters.emzed_store_url = "http://uweschmitt.info:3141/root/dev"
        self.parameters.emzed_store_index_url = "http://uweschmitt.info:3141/root/dev/+simple/"
        self.parameters.pypi_url = "http://testpypi.python.org/pypi"

global_config = _UserConfig()
