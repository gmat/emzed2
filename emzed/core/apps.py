# encoding:latin-1

EMZED_APP_MARKER_FILE = ".emzed_app_marker"

import os
import requests
import config
import subprocess


SETUP_PY_TEMPALTE = """
import distutils.config

def patched(self):
    return dict(realm="pypi",
                username=%(user)r,
                password=%(password)r,
                repository=%(repository)r,
                server="local",
                )
distutils.config.PyPIRCCommand._read_pypirc = patched

from setuptools import setup

setup(name=%(app_name)r,
      packages=[ %(app_name)r],
      version=%(version)r,
      author=%(author)r,
      author_email=%(author_email)r,
      url=%(author_url)r,
      entry_points = {
          'emzed_plugin' :
          [
              "package = %(app_name)s",
              ]
          }
     )
"""

def _normalize(folder):
    return os.path.abspath(os.path.normpath(folder))

def _check_name(app_name):
    forbidden = " .-"
    if any(c in app_name for c in forbidden) or app_name.lower() != app_name:
        suggested = app_name.lower()
        for f in forbidden:
            suggested = suggested.replace(f, "_")
        raise Exception("name %r invalid. you could use %r instead" % (
            app_name, suggested))


def _test_if_folder_is_inside_existing_app(folder):
    folder = _normalize(folder)
    # normpath replaces "/" to r"\" on win:
    for __ in range(100):
        if folder.endswith(os.path.sep):
            return None # reached top folder (maybe something like C:\ on win)
        # go one folder level upwards:
        folder, __ = os.path.split(folder)
        if os.path.exists(os.path.join(folder, EMZED_APP_MARKER_FILE)):
                return folder
    raise Exception("unlimited loop for folder %r" % folder)


def _test_if_folder_already_exists(folder):
    if os.path.exists(folder):
        raise Exception("%s already exists" % folder)


def _create_app_folder(app_folder, app_name):
    path_to_existing_app = _test_if_folder_is_inside_existing_app(app_folder)
    if path_to_existing_app:
        raise Exception("found existing app in %s" % path_to_existing_app)

    os.makedirs(app_folder)
    open(os.path.join(app_folder, EMZED_APP_MARKER_FILE), "w").close()
    _create_package_folder(app_folder, app_name)
    _create_test_folder(app_folder, app_name)


def _create_package_folder(app_folder, app_name):
    package_folder = os.path.join(app_folder, app_name)
    os.makedirs(package_folder)
    with open(os.path.join(package_folder, "__init__.py"), "w") as fp:
        fp.write("""
from hello import hello
    """)

    with open(os.path.join(package_folder, "hello.py"), "w") as fp:
        fp.write("""
def hello(): 
    return "this is %s"
    """ % app_name)

    with open(os.path.join(app_folder, "setup.py"), "w") as fp:
        user = config.get_value("app_store", "user")
        password = config.get_value("app_store", "password")
        repository = config.get_url("app_store", "app_store_url") + "/"
        author = config.get_value("app_store", "author")
        author_email = config.get_value("app_store", "author_email")
        author_url = config.get_url("app_store", "author_url")
        version = "0.0.1"
        fp.write(SETUP_PY_TEMPALTE % locals())


def _create_test_folder(app_folder, app_name):
    tests_folder = os.path.join(app_folder, "tests")
    os.makedirs(tests_folder)
    with open(os.path.join(tests_folder, "__init__.py"), "w") as fp:
        pass

    with open(os.path.join(tests_folder, "test_hello.py"), "w") as fp:
        fp.write("""
import %(app_name)s
def test_hello():
    assert isinstance(%(app_name)s.hello(), basestring), %(app_name)s.hello()
    """ % locals())


def create_app_scaffold(folder, app_name):
    _check_name(app_name)
    folder = _normalize(folder)
    _test_if_folder_already_exists(folder)
    _create_app_folder(folder, app_name)

def delete_from_app_store(app_name):
    user = config.get_value("app_store", "user")
    password = config.get_value("app_store", "password")
    app_url = "%s/%s" % (config.get_url("app_store", "app_store_url"), app_name)
    response = requests.delete(app_url, auth=(user, password))
    response.raise_for_status()

def upload_to_app_store(app_folder):
    os.chdir(app_folder)
    rc = subprocess.call("python setup.py sdist upload", shell=True)
    if rc:
        raise Exception("upload failed")

def install_from_app_store(app_name):
    index_url = config.get_url("app_store", "app_store_index_url")
    exit_code = subprocess.call("pip install -i %s %s" % (index_url, app_name),
        shell=True)
    assert exit_code == 0

def uninstall_app(app_name):
    exit_code = subprocess.call("pip uninstall --yes %s" % app_name,
        shell=True)
    assert exit_code == 0
    import emzed.ext
    delattr(emzed.ext, app_name)

