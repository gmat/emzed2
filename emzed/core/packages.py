# encoding:latin-1

EMZED_PKG_MARKER_FILE = ".emzed_pkg_marker"

import os
import sys
import requests
import subprocess
import pkg_resources

import helpers
from config import global_config
import licenses


SETUP_PY_TEMPLATE = """

# YOU CAN EDIT THESE FIELDS ######################################################################

# install package as emzed extension ?
# -> package will appear in emzed.ext namespace after installation

IS_EXTENSION = True

# Install package as emzed app ?
# will can be started as app.%(pkg_name)s.run()

# set this variable to None if this is a pure extension and not an emzed app
APP_MAIN = "%(pkg_name)s.main:run"

VERSION = %(version)r
AUTHOR = %(author)r
AUTHOR_EMAIL = %(author_email)r
AUTHOR_URL = %(author_url)r

PKG_NAME = %(pkg_name)r

DESCRIPTION = "please describe here %(pkg_name)s in one line"
LONG_DESCRIPTION = \"\"\"

describe %(pkg_name)s here in more than one line

\"\"\"

LICENSE = "http://opensource.org/licenses/GPL-3.0"

# DO NOT TOUCH THE CODE BELOW UNLESS YOU KNOW WHAT YOU DO !!!!  # ################################

if APP_MAIN is not None:
    try:
        mod_name, fun_name = APP_MAIN.split(":")
        exec "import %%s as _mod" %% mod_name
        fun = getattr(_mod, fun_name)
    except:
        raise Exception("invalid specification %%r of APP_MAIN" %% APP_MAIN)


entry_points = dict()
entry_points['emzed_package'] = [ "package = " + PKG_NAME, ]
if IS_EXTENSION:
    entry_points['emzed_package'].append("extension = " + PKG_NAME)
if APP_MAIN is not None:
    entry_points['emzed_package'].append("main = %%s" %% APP_MAIN)


if __name__ == "__main__":   # allows import setup.py for version checking

    import distutils.config

    def patched(self):
        return dict(realm="pypi",
                    username=%(user)r,
                    password=%(password)r,
                    repository=%(repository)r,
                    server="local",
                    )
    distutils.config.PyPIRCCommand._read_pypirc = patched


    from setuptools import setup
    setup(name=PKG_NAME,
        packages=[ PKG_NAME ],
        version=".".join(map(str, VERSION)),
        author=AUTHOR,
        author_email=AUTHOR_EMAIL,
        url=AUTHOR_URL,
        description=DESCRIPTION,
        long_description=LONG_DESCRIPTION,
        license=LICENSE,
        entry_points = entry_points
        )
   """

def _normalize(folder):
    return os.path.abspath(os.path.normpath(folder))

def _check_name(pkg_name):
    forbidden = " .-"
    if any(c in pkg_name for c in forbidden) or pkg_name.lower() != pkg_name:
        suggested = pkg_name.lower()
        for f in forbidden:
            suggested = suggested.replace(f, "_")
        raise Exception("name %r invalid. you could use %r instead" % ( pkg_name, suggested))


def _test_if_folder_is_inside_existing_pkg(folder):
    folder = _normalize(folder)
    # normpath replaces "/" to r"\" on win:
    for __ in range(100):
        if folder.endswith(os.path.sep):
            return None # reached top folder (maybe something like C:\ on win)
        # go one folder level upwards:
        folder, __ = os.path.split(folder)
        if os.path.exists(os.path.join(folder, EMZED_PKG_MARKER_FILE)):
                return folder
    raise Exception("unlimited loop for folder %r" % folder)


def _test_if_folder_already_exists(folder):
    if os.path.exists(folder):
        raise Exception("%s already exists" % folder)


def _create_pkg_folder(pkg_folder, pkg_name, version):
    path_to_existing_pkg = _test_if_folder_is_inside_existing_pkg(pkg_folder)
    if path_to_existing_pkg:
        raise Exception("found existing emzed package in %s" % path_to_existing_pkg)

    os.makedirs(pkg_folder)
    open(os.path.join(pkg_folder, EMZED_PKG_MARKER_FILE), "w").close()
    _create_package_folder(pkg_folder, pkg_name, version)
    _create_test_folder(pkg_folder, pkg_name)


def _create_package_folder(pkg_folder, pkg_name, version):
    package_folder = os.path.join(pkg_folder, pkg_name)
    os.makedirs(package_folder)
    with open(os.path.join(package_folder, "__init__.py"), "w") as fp:
        fp.write("""
from hello import hello
    """)

    with open(os.path.join(package_folder, "main.py"), "w") as fp:
        fp.write("""
def run():
    return 42
    """)

    with open(os.path.join(package_folder, "hello.py"), "w") as fp:
        fp.write("""
def hello():
    return "%s says hello"
    """ % pkg_name)

    with open(os.path.join(pkg_folder, "setup.py"), "w") as fp:
        user = global_config.get("emzed_store_user")
        password = global_config.get("emzed_store_password")
        repository = global_config.get_url("emzed_store_url")
        author = global_config.get("user_name")
        author_email = global_config.get("user_email")
        author_url = global_config.get("user_url")
        fp.write(SETUP_PY_TEMPLATE % locals())

    with open(os.path.join(pkg_folder, "LICENSE"), "w") as fp:
        fp.write(licenses.GPL_3)

    with open(os.path.join(pkg_folder, "README"), "w") as fp:
        fp.write("please describe your emzed package here\n")

def _create_test_folder(pkg_folder, pkg_name):
    tests_folder = os.path.join(pkg_folder, "tests")
    os.makedirs(tests_folder)
    with open(os.path.join(tests_folder, "__init__.py"), "w") as fp:
        pass

    with open(os.path.join(tests_folder, "test_main.py"), "w") as fp:
        fp.write("""
import %(pkg_name)s.main
def test_hello():
    assert isinstance(%(pkg_name)s.main.run(), basestring)
    """ % locals())


def create_package_scaffold(folder, pkg_name, version=(0,0,1)):
    _check_name(pkg_name)
    folder = _normalize(folder)
    _test_if_folder_already_exists(folder)
    _create_pkg_folder(folder, pkg_name, version)

def delete_from_emzed_store(pkg_name):
    user = global_config.get("emzed_store_user")
    password = global_config.get("emzed_store_password")
    url = global_config.get_url("emzed_store_url") + pkg_name
    response = requests.delete(url, auth=(user, password))
    response.raise_for_status()

def upload_to_emzed_store(pkg_folder):
    os.chdir(pkg_folder)

    # make sure we load the right setup.py
    sys.path.insert(0, os.path.abspath(pkg_folder))
    import setup
    sys.path.pop(0)

    for p, versions in list_packages_from_emzed_store():
        if p == setup.PKG_NAME and setup.VERSION  in versions:
            raise Exception("package %s with version %s already exists" % (setup.PKG_NAME,
                                                                           setup.VERSION))

    rc = subprocess.call("python setup.py sdist upload", shell=True)
    if rc:
        raise Exception("upload failed")

def install_from_emzed_store(pkg_name, version=None):
    if version:
        assert isinstance(version, tuple)
        assert len(version) == 3
    index_url = global_config.get_url("emzed_store_index_url")
    pkg_query = pkg_name
    if version:
        pkg_query += "==%s.%s.%s" % version

    is_venv = os.environ.get("VIRTUAL_ENV") is not None
    user_flag = "" if is_venv else "--user"

    exit_code = subprocess.call("pip install %s -i %s %s" % (user_flag, index_url, pkg_query),
                                shell=True)
    assert exit_code == 0

def uninstall_emzed_package(pkg_name):
    exit_code = subprocess.call("pip uninstall --yes %s" % pkg_name, shell=True)
    assert exit_code == 0
    import emzed.ext
    delattr(emzed.ext, pkg_name)


def installed_emzed_packages():
    reload(pkg_resources)
    entry_points = pkg_resources.iter_entry_points
    packages = [ep.module_name for ep in entry_points("emzed_package", name="package")]
    extensions = [ep.module_name for ep in entry_points("emzed_package", name="extension")]
    apps = [ep.module_name for ep in entry_points("emzed_package", name="main")]

    return [ (p, p in extensions, p in apps) for p in packages]

def list_packages_from_emzed_store():
    url = global_config.get_url("emzed_store_url")
    response = helpers.get_json(url)
    response.raise_for_status()
    packages = [ name.encode("latin-1") for name in response.json()["result"]]
    result = []
    for package in packages:
        response = helpers.get_json(url + package)
        response.raise_for_status()
        version_strings = response.json()["result"].keys()
        versions = [map(int, version_str.split(".")) for version_str in version_strings]
        result.append((package, map(tuple, versions)))
    return result

def list_newest_packages_from_emzed_store():
    packages = list_packages_from_emzed_store()
    return [ (pkg, max(versions)) for (pkg, versions) in packages]
