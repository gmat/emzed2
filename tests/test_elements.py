from emzed.core.chemistry import Elements, MonoIsotopicElements
from emzed.core.chemistry import monoisotopicMass

import pyopenms

IS_PYOPENMS_2 = pyopenms.__version__.startswith("2.")


def test_elements(regtest):
    el = MonoIsotopicElements()
    el2 = MonoIsotopicElements()
    assert el.rows is el2.rows  # check if borg is working. which reduces startup

    el.sortBy("number")
    el.print_(out=regtest)

    el3 = Elements()
    el3.sortBy("number")

    assert len(el3) == 199 if IS_PYOPENMS_2 else 102

def test_mass_calculation():

    el = MonoIsotopicElements()
    m0H = monoisotopicMass("H")
    assert abs(m0H - el.m0.values[1]) < 1e-12
    m0H = monoisotopicMass("H2")
    assert abs(m0H - 2 * el.m0.values[1]) < 1e-12

    assert abs(monoisotopicMass("NaCl") - 57.9586219609) < 1e-7, monoisotopicMass("NaCl")

