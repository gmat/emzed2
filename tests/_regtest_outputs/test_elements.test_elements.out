number   symbol   name       massnumber m0
int      str      str        int        float
------   ------   ------     ------     ------
1        D        Deuterium  1          2.0141017800
1        H        Hydrogen   1          1.0078250319
1        T        Tritium    1          3.0160492700
2        He       Helium     3          3.0160293191
3        Li       Lithium    6          6.0151220000
4        Be       Beryllium  9          9.0121822000
5        B        Bor        10         10.0129370000
6        C        Carbon     12         12.0000000000
7        N        Nitrogen   14         14.0030740000
8        O        Oxygen     16         15.9949150000
9        F        Fluorine   19         18.9984032200
11       Na       Sodium     23         22.9897692809
12       Mg       Magnesium  24         23.9850420000
13       Al       Aluminium  27         26.9815386300
14       Si       Silicon    28         27.9769265325
15       P        Phosphorus 31         30.9737614900
16       S        Sulfur     32         31.9720707300
17       Cl       Chlorine   35         34.9688526800
18       Ar       Argon      36         35.9675451060
19       K        Potassium  39         38.9637066800
20       Ca       Calcium    40         39.9625909800
22       Ti       Titanium   46         45.9526316000
23       V        Vanadium   50         49.9471585000
24       Cr       Chromium   50         49.9460442000
25       Mn       Manganese  55         54.9380500000
26       Fe       Ferrum     54         53.9396105000
27       Co       Cobalt     59         58.9331950000
28       Ni       Nickel     58         57.9353480000
29       Cu       Copper     63         62.9296010000
30       Zn       Zinc       64         63.9291470000
31       Ga       Gallium    69         68.9255736000
32       Ge       Germanium  70         69.9242474000
33       As       Arsenic    75         74.9215965000
34       Se       Selenium   74         73.9224764000
35       Br       Bromine    79         78.9183371000
37       Rb       Rubidium   85         84.9117897380
38       Sr       Strontium  84         83.9134250000
40       Zr       Zirconium  90         89.9047044000
42       Mo       Molybdenum 92         91.9068100000
44       Ru       Ruthenium  96         95.9075980000
46       Pd       Palladium  102        101.9056090000
47       Ag       Silver     107        106.9050930000
48       Cd       Cadmium    106        105.9064580000
50       Sn       Tin        112        111.9048180000
51       Sb       Antimony   121        120.9038157000
52       Te       Tellurium  120        119.9040200000
53       I        Iodine     127        126.9044730000
55       Cs       Caesium    133        132.9054519330
56       Ba       Barium     132        131.9050613000
58       Ce       Cerium     136        135.9071720000
64       Gd       Gadolinium 152        151.9197910000
72       Hf       Hafnium    176        175.9414086000
73       Ta       Tantalum   181        180.9479958000
74       W        Tungsten   180        179.9467040000
78       Pt       Platinum   192        191.9610380000
79       Au       Gold       197        196.9665510000
80       Hg       Mercury    196        195.9658330000
81       Tl       Thallium   203        202.9723442000
82       Pb       Lead       204        203.9730436000
83       Bi       Bismuth    209        208.9803987000
